pkg/nodeadm/nodeadm.go

```diff
@@ -164,44 +121,48 @@ func setInitConfigFromClusterConfig(cfg *InitConfiguration, clusterConfig *spv1.
 		cfg.MasterConfiguration.KubeProxy = *clusterConfig.KubeProxy
 	}
 	cfg.MasterConfiguration.SchedulerExtraArgs = clusterConfig.KubeScheduler
-	cfg.Kubelet = clusterConfig.Kubelet
+	if clusterConfig.Kubelet != nil {
+		cfg.MasterConfiguration.KubeletConfiguration = *clusterConfig.Kubelet
+	}
 	cfg.NetworkBackend = clusterConfig.NetworkBackend
 	cfg.KeepAlived = clusterConfig.KeepAlived
 	return nil
 }
 
-func setJoinConfigFromClusterConfig(cfg *JoinConfiguration, clusterConfig *spv1.ClusterConfig) {
-	cfg.Kubelet = clusterConfig.Kubelet
-}
+func JoinConfigurationForMachine(cluster *clusterv1.Cluster, machine *clusterv1.Machine, discoveryTokenAPIServers, discoveryTokenCACertHashes []string, token string) (*JoinConfiguration, error) {
+	cfg := &JoinConfiguration{
+		NodeConfiguration: KubeadmNodeConfiguration{
+			TypeMeta: metav1.TypeMeta{
+				Kind:       "NodeConfiguration",
+				APIVersion: "kubeadm.k8s.io/v1alpha2",
+			},
+		},
+	}
 
-func JoinConfigurationForMachine(cluster *clusterv1.Cluster, machine *clusterv1.Machine) (*JoinConfiguration, error) {
-	cfg := &JoinConfiguration{}
+	// Discovery and TLS bootstrap configuration
+	cfg.NodeConfiguration.DiscoveryTokenAPIServers = discoveryTokenAPIServers
+	cfg.NodeConfiguration.DiscoveryTokenCACertHashes = discoveryTokenCACertHashes
+	cfg.NodeConfiguration.Token = token
 
-	cpc, err := controller.GetClusterSpec(*cluster)
-	if err != nil {
-		return nil, fmt.Errorf("unable to decode cluster spec: %v", err)
-	}
+	// NodeRegistrationOptions
+	cfg.NodeConfiguration.NodeRegistration.Name = machine.Name
+	cfg.NodeConfiguration.NodeRegistration.Taints = machine.Spec.Taints
 
-	// Networking
-	switch len(cluster.Spec.ClusterNetwork.Pods.CIDRBlocks) {
-	case 0:
-		return nil, fmt.Errorf("cluster %q spec.clusterNetwork.pods.cidrBlocks must contain at least one block", cluster.Name)
-	case 1:
-		cfg.Networking.PodSubnet = cluster.Spec.ClusterNetwork.Pods.CIDRBlocks[0]
-	default:
-		return nil, fmt.Errorf("cluster %q spec.clusterNetwork.pods.cidrBlocks must contain at most one block", cluster.Name)
-	}
-	switch len(cluster.Spec.ClusterNetwork.Services.CIDRBlocks) {
-	case 0:
-		return nil, fmt.Errorf("cluster %q spec.clusterNetwork.pods.cidrBlocks must contain at least one block", cluster.Name)
-	case 1:
-		cfg.Networking.ServiceSubnet = cluster.Spec.ClusterNetwork.Services.CIDRBlocks[0]
-	default:
-		return nil, fmt.Errorf("cluster %q spec.clusterNetwork.pods.cidrBlocks must contain at most one block", cluster.Name)
+	return cfg, nil
+}
+
+func validateClusterNetworkingConfiguration(cluster clusterv1.Cluster) error {
+	switch cbl := len(cluster.Spec.ClusterNetwork.Pods.CIDRBlocks); {
+	case cbl < 1:
```

Comment:
why not just cbl != 1 here?

Reply:
I think the error is more helpful when it is specific ("at most one" or "at least one").

If I want the more helpful error and I use `cbl != 1` here, I'll still have to handle the two cases separately.